quadriga/src/main/java/edu/asu/diging/quadriga/core/citesphere/impl/CitesphereConnector.java

```diff
@@ -0,0 +1,135 @@
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.Scope;
+import com.nimbusds.oauth2.sdk.TokenRequest;
+import com.nimbusds.oauth2.sdk.TokenResponse;
+import com.nimbusds.oauth2.sdk.auth.ClientAuthentication;
+import com.nimbusds.oauth2.sdk.auth.ClientSecretBasic;
+import com.nimbusds.oauth2.sdk.auth.Secret;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.oauth2.sdk.token.AccessToken;
+
+import edu.asu.diging.quadriga.core.citesphere.ICitesphereConnector;
+import edu.asu.diging.quadriga.core.model.citesphere.CitesphereAppInfo;
+
+@Service
+@PropertySource("classpath:/config.properties")
+public class CitesphereConnector implements ICitesphereConnector {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Value("${citesphere_client_id}")
+    private String citesphereClientId;
+
+    @Value("${citesphere_client_secret}")
+    private String citesphereSecret;
+
+    @Value("${citesphere_base_url}")
+    private String citesphereBaseUrl;
+
+    @Value("${citesphere_token_endpoint}")
+    private String citesphereTokenEndpoint;
+
+    @Value("${citesphere_check_token_endpoint}")
+    private String citesphereCheckTokenEndpoint;
+
+    @Value("${citesphere_scopes}")
+    private String citesphereScopes;
+
+    @Value("${citesphere_get_apps_endpoint}")
+    private String citesphereAppsEndpoint;
+
+    private RestTemplate restTemplate;
+
+    private String currentAccessToken;
+
+    public CitesphereConnector() {
+        restTemplate = new RestTemplate();
+    }
+
+    /* (non-Javadoc)
+     * @see edu.asu.diging.quadriga.core.citesphere.ICitesphereConnector#getCitesphereApps()
+     */
+    @Override
+    public List<CitesphereAppInfo> getCitesphereApps() throws HttpClientErrorException {
+        if (currentAccessToken == null) {
+            currentAccessToken = getAccessToken();
+        }
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("Authorization", "Bearer " + currentAccessToken);
+        HttpEntity<String> entity = new HttpEntity<String>(headers);
+
+        String appUrl = citesphereBaseUrl + citesphereAppsEndpoint;
+
+        try {
+            return restTemplate.exchange(appUrl, HttpMethod.GET, entity,
+                    new ParameterizedTypeReference<List<CitesphereAppInfo>>() {
+                    }).getBody();
+        } catch (HttpClientErrorException ex) {
+            if (ex.getStatusCode() == HttpStatus.UNAUTHORIZED) {
+                currentAccessToken = getAccessToken();
+                headers.set("Authorization", "Bearer " + currentAccessToken);
+                return restTemplate.exchange(appUrl, HttpMethod.GET, entity,
+                        new ParameterizedTypeReference<List<CitesphereAppInfo>>() {
+                        }).getBody();
+            }
+            throw ex;
+        }
+    }
+
+    private String getAccessToken() {
+        AuthorizationGrant clientGrant = new ClientCredentialsGrant();
```

Comment:
I have written the exact same method (branch story/Q20-3) in a 'TokenValidatorImpl'
Since you're writing that same method here, should I remove it from my changes and use this method instead?
In that case, you might have to make this method public

Reply:
I will make it public then.

Reply:
@Pooja444, @MaulikLim depending on what the `TokenValidatorImpl` does the classes should probably be better encapsulated. Since `getCitesphereApps` is implemented here (using this method) and there is another method in a different class that also uses `getAccessToken` it sounds like you either want all methods connecting to citesphere (using `getAccessToken`) in this class or have low level methods like `getAccessToken` in it's own class. Does that make sense?

Reply:
Yes it does. We'll discuss and make the changes.